File: go/plugins/plugins.go
package plugins

import (
	plugin "github.com/voxgig/plugin/go/plugin"

	"github.com/voxgig/sekreto/go/plugins/aws"
	"github.com/voxgig/sekreto/go/plugins/azuresecrets"
	"github.com/voxgig/sekreto/go/plugins/boru"
	"github.com/voxgig/sekreto/go/plugins/doppler"
	"github.com/voxgig/sekreto/go/plugins/gcpsecrets"
	"github.com/voxgig/sekreto/go/plugins/hashicorp"
	"github.com/voxgig/sekreto/go/plugins/infisical"
	"github.com/voxgig/sekreto/go/plugins/minivault"
	"github.com/voxgig/sekreto/go/plugins/onepassword"
	"github.com/voxgig/sekreto/go/plugins/secretspec"
)

// All is every plugin definition this library ships, in a fresh slice.
func All() []plugin.Definition {
	return []plugin.Definition{
		hashicorp.Plugin, boru.Plugin, aws.Secrets, aws.Params, gcpsecrets.Plugin,
		azuresecrets.Plugin, onepassword.Plugin, doppler.Plugin, infisical.Plugin,
		secretspec.Plugin, minivault.Plugin,
	}
}

File: php/plugins/plugins.php
<?php

/**
 * THE FULL SET - every plugin kind this library ships, in one call.
 *
 * One plugin is its own file, and requiring it requires nothing else:
 *
 *     require_once __DIR__ . '/plugins/hashicorp.php';
 *
 *     use Voxgig\Sekreto\Sekreto;
 *     use function Voxgig\Sekreto\Plugins\hashicorp;
 *
 *     $secrets = new Sekreto([
 *         'plugins' => [hashicorp()],
 *         'providers' => [['kind' => 'env'], ['kind' => 'hashicorp', ...]],
 *     ]);
 *
 * THIS FILE IS THE ONE THING THAT REQUIRES ALL TEN. It is for the CLI, the
 * conformance suite, and an app whose chain is decided at run time.
 * Reaching it pulls in every network client, AWS request signing and the
 * two child-process kinds - which is the cost the core/plugin split exists
 * to remove, so an app requires the kinds it actually configures, each
 * from its own file.
 *
 * The requires are INSIDE the function on purpose. At file scope they
 * would make requiring this file - to read the list, to name the type -
 * load all ten as a side effect, and a side effect of an import is exactly
 * what the split forbids (docs/design/plugin-providers.md).
 */

declare(strict_types=1);

namespace Voxgig\Sekreto\Plugins;

/**
 * Every plugin kind, built on demand, in the order the design doc lists
 * them.
 *
 * @return array<int, array<string, mixed>>
 */
function allplugins(): array
{
    require_once __DIR__ . '/hashicorp.php';
    require_once __DIR__ . '/boru.php';
    require_once __DIR__ . '/aws.php';
    require_once __DIR__ . '/gcpsecrets.php';
    require_once __DIR__ . '/azuresecrets.php';
    require_once __DIR__ . '/onepassword.php';
    require_once __DIR__ . '/doppler.php';
    require_once __DIR__ . '/infisical.php';
    require_once __DIR__ . '/secretspec.php';

    return [
        hashicorp(), boru(), awssecrets(), awsparams(), gcpsecrets(),
        azuresecrets(), onepassword(), doppler(), infisical(), secretspec(),
    ];
}
